if (interactive()) { options(error=recover) }

printbad <- function(...) { print(paste0(...)) }

ARG_DELIM=','

require("optparse")
require("edgeR")

option_list <- list(
     make_option(c("-v", "--verbose"), action="store_true", default=FALSE
                ,help="Print extra-verbose output")
   , make_option(c("-g", "--groups"), type="character", default=NULL
                ,help="REQUIRED. Comma-delimited GROUPS that each sample belongs to. Example: --groups='1,1,1,2,2,2' .")
   , make_option(c("-i", "--countfiles"), type="character", default=NULL
                ,help="REQUIRED. Specify a comma-separated list of input COUNT matrixes. Assumed to have **NO HEADERS** and TWO valid columns, as follows: 1) Identifier of a gene/feature, and 2) expression level (counts). The expectation is that these various files will be generated by htseq-counts. If you use edgeR (or something else sensible that makes a single matrix file), then this code will require some modification.")
   , make_option(c("-o", "--outfile"), type="character", default=NULL
                ,help="REQUIRED. Specify a (single) edgeR output table file.")
)
opt <- parse_args(OptionParser(option_list=option_list))



grvec <- strsplit(opt$groups, split=ARG_DELIM)[[1]]
# Should probably all be numeric. Definitely should be same length as number of columns in input matrix

matvec <- strsplit(opt$groups, split=ARG_DELIM)[[1]]

if (!all(file.exists(matvec))) {
     print("Failure -- missing input matrix file. Here are the files, and whether or not they were present:")
     print(matvec)
     print(file.exists(matvec))
     stop(paste0("Fatal error--one of the expected input matrix files appears NOT to exist!"))
}


agw_handle_rna_diff_expression(
     count_filenames.vec=opt$countfiles # assumed to be from HTSEQ-counts
   , outfile=opt$outfile
   , groups.vec=grvec)

agw_de_two_groups <- function(theRG, g1, g2, theGroupVec, writeOutputToThisFile=FALSE) {
     # g1 and g2 are both NUMERIC
     stopifnot(g1 != g2)
     stopifnot(g1 >= 1);
     stopifnot(g2 >= 1);
     D <- DGEList(counts=theRG, group=theGroupVec)
     D <- edgeR::estimateCommonDisp(D)       
     D <- edgeR::estimateTagwiseDisp(D) # exactTest for negative bionomial distribution
     groupCounts        <- table(theGroupVec)
     bothHaveReplicates <- (groupCounts[[g1]] >= 2 && groupCounts[[g2]] >= 2)
     if (bothHaveReplicates) {
          et <- edgeR::exactTest(D, pair=c(g1, g2))
     } else {
          HARD_CODED_DISPERSION <- 0.1 # may be low for human (0.4 seems recommended?). 0.1 is recommended for mouse and other genetically-identical samples though
          # Note about HARD_CODED_DISPERSION: this number was used by Stacia during initial analysis, so we need to keep it the same here too!
          et <- edgeR::exactTest(D, pair=c(g1, g2), dispersion=HARD_CODED_DISPERSION)
     }
     tt <- data.frame(topTags(et, n=nrow(et), sort.by="none"), check.names=FALSE) # gets the FDR calculated, too
     t.df <- tt[, c("logFC", "PValue", "FDR")] # only the three columns of interest---omit the logCPM column
     colnames(t.df) <- paste( paste0(g1,":",g2), colnames(t.df), sep=" ")
     
     if (writeOutputToThisFile) {
          outTable <- data.frame("IDENTIFIER"=rownames(t.df), t.df, check.names=FALSE)
          print0("Writing to the output file <", writeOutputToThisFile, ">")
          write.table(outTable, file=writeOutputToThisFile, sep="\t", quote=FALSE, col.names=TRUE, row.names=FALSE)
     }
     return(t.df) # Return the table with logFC, PValue, FDR. It's a data frame now, instead of whatever topTags natively returns.
}

agw_handle_rna_diff_expression <- function(count_filenames.vec, outfile, groups.vec) {
     
     smallest_num_replicates <- min( table(groups.vec) )
     if (smallest_num_replicates < 2) {
	  warning("No group has more than one replicate, skip this comparison! Maybe we should just report fold change (?).") # TODO: maybe report just fold change?
          stop("No group has more than one replicate, skip this comparison! Maybe we should just report fold change (?).")
     } else if (!all(sapply(count_filenames.vec, file.nonzero.exists))) {
	  stop("EDGER IS MISSING SOME HTSEQ INPUT FILES for that experiment ID so we are skipping this differential expression.")
     }
     
     print("Running EdgeR...")
     stopifnot(all(sapply(count_filenames.vec, file.nonzero.exists))) # all the input files must ALREADY exist, please!
     RG <- edgeR::readDGE(count_filenames.vec, header=FALSE, group=groups.vec) # htseq-count does NOT have a header!
     # """Each file is assumed to contain digital gene expression data for
     # one genomic sample or count library, with gene identifiers in the
     # first column and counts in the second column. Gene identifiers are
     # assumed to be unique and not repeated in any one file.  The
     # A count of zero will be entered for any
     # gene that was not found in any particular sample.
     #By default, the files are assumed to be tab-delimited and to
     # contain column headings. Other file formats can be handled by..."""
     
     if (ncol(RG) != length(groups.vec)) {
          stop("The number of 'count matrix columns is NOT equal to the length of the group vector. This probably means some experimental groups were not aligned and/or processed downstream.")
     }
     
     # If there are replicates:
     # Figure out groups for pairwise analysis
     stopifnot(is.integer(groups.vec))
     if (any(sort(unique(groups.vec)) != seq_along(unique(groups.vec)))) {
          stop("ERROR 162F: the groups.vec were NOT numbered 1-through-n with no gaps! This indicates something we did not expect in the input data.")
     }
     
     elist <- list()
     for (g1 in sort(unique(groups.vec))) { # g1 and g2 are both NUMERIC GROUP IDs
          for (g2 in sort(unique(groups.vec))) {
               if (g1 >= g2) { next; } # don't double-calculate, or calculate g1 vs itself
               print(paste0("Comparing group ", g1, " vs ", g2, " (comparison name is ", g1, ":", g2, ")..."))
               elist[[ 1+length(elist) ]] <- agw_de_two_groups(theRG=RG, g1=g1, g2=g2, theGroupVec=groups.vec)
          }
     }
     
     #topTags(et)
     # If there are no replicates:
     # D <- DGEList(counts=RG,group=1:2:3:4)
     # et <- exactTest(D,dispersion=0.1)
     ## topTags(et)
     
     if (length(elist) > 0) {
          eframe <- do.call(cbind, elist)
          outTable <- data.frame("IDENTIFIER"=rownames(elist[[1]]), eframe, check.names=FALSE)
          write.table(outTable, file=outfile, sep="\t", quote=FALSE, col.names=TRUE, row.names=FALSE)
          print(paste0("Wrote results to the output file <", outfile, ">"))
     } else {
          outDat <- paste0(outfile, "--FAILED-TO-GENERATE-RESULTS.txt")
          write.table(c("failure - no results generated by the edgeR code"), file=outfile)
          printbad("Warning: did not find any group comparisons for this experiment! This may be a programming error (?) or it may be due to unusual input data.")
     }
     
}
